### PR TITLE
[TIMOB-26538] Android: Modified services/broadcast-receivers to be terminated upon app exit

### PR DESCRIPTION
**Pull request by jquick-axway**
**JIRA:**
- https://jira.appcelerator.org/browse/TIMOB-26538
- https://jira.appcelerator.org/browse/TIMOB-26555

**Summary:**
- Fixes bug where starting a service/broadcast-receiver, backing-out of the app, and then restarting the app would cause the app to be stuck at the splash screen.
  * This happened because only one Titanium JS runtime can be ran at a time.
  * This PR partially reverts the change made by #2658 .
  * Better solution would be to set root window's "exitOnClose" property to `false`. This allows UI, service, and JS runtime to persist when backing out of the root window and makes it work more like iOS.
- Added new `OnDisposingListener` to Java `KrollRuntime` class.
  * In the future, we can use this to add support for `Ti.App` "close" event property on Android.
- This change is needed to avoid tech-support issues since foreground services are becoming increasingly popular on Android.

**Broadcast Receiver Test:**
1. Build and run the code attached to [TIMOB-26555](https://jira.appcelerator.org/browse/TIMOB-26555) on Android.
2. Press the back button to exit the app.
3. Relaunch the app.
4. Verify that the app is not stuck on the splash screen.

**Service Test:**
1. Set up a classic Titanium app with the below.
2. Build and run on an Android device.
3. Verify on app startup that a notification was added to the status bar. _(This was added by the 
service.)_
4. Press the back button to exit the app.
5. Verify that the app's notification was removed. _(This indicates that the service was stopped)_
6. Relaunch the app.
7. Verify that a notification was re-added to the status bar. _(This is the bug fix.)_
8. In the "app.js", uncomment the "exitOnClose" property.
9. Build and run on Android.
10. Verify that notification was posted on app startup.
11. Exit out of the app via the back button.
12. Verify that the notification is still in the status bar.
13. Relaunch the app.
14. Verify that the splash screen was **NOT** shown when launched. _(This is because "exitOnClose" `false` does not destroy the main window. It is resuming instead like iOS.)_

---
tiapp.xml
```xml
<?xml version="1.0" encoding="UTF-8"?>
<ti:app xmlns:ti="http://ti.appcelerator.org">
	<android xmlns:android="http://schemas.android.com/apk/res/android">
		<services>
			<service url="Service.js"/>
		</services>
	</android>
</ti:app>
```

---
./Resources/Service.js
```javascript
Ti.API.info("@@@ Service started.");

// Gets an intent used to resume the app if running in the background.
function getAppResumeIntent() {
	var intent = Ti.App.Android.launchIntent;
	if (!intent) {
		intent = Ti.Android.createIntent({});
	}
	return intent;
}

// Put this service into the foreground state.
var notificationId = 123;
Ti.Android.currentService.foregroundNotify(notificationId, Ti.Android.createNotification({
	contentTitle: "Foreground Service",
	contentText: "Content Text",
	contentIntent: Ti.Android.createPendingIntent({
		// Using the launch intent will resume the app when tapped.
		intent: getAppResumeIntent(),
	}),
}));

// Log a message every 1 second.
setInterval(function() {
	Ti.API.info("@@@ Service timer elapsed. " + (new Date()).toString());
}, 1000);
```

---
./Resources/app.js
```javascript
// Start the service.
var intent = Ti.Android.createServiceIntent({
	url: "Service.js",
});
Ti.Android.startService(intent);

// Display the main window.
var window = Ti.UI.createWindow({
//	exitOnClose: false,
});
window.open();
```

### CODE REVIEW

**Comment by lokeshchdhry** [comment]
FR Passed.

`services/broadcast receivers` are successfully terminated on app close. If `exitOnClose: false` then they are not terminated.

Studio Ver:     5.1.1.201809051655
SDK Ver:        8.0.0 local build
OS Ver:         10.14
Xcode Ver:      Xcode 10.1
Appc NPM:       4.2.13
Appc CLI:       7.0.8-5
Daemon Ver:     1.1.3
Ti CLI Ver:     5.1.1
Alloy Ver:      1.13.4
Node Ver:       8.9.1
NPM Ver:        5.5.1
Java Ver:       10.0.2
Devices:        ⇨ google Nexus 5 (Android 6.0.1)
                ⇨ google Nexus 6P (Android 8.1.0)

**Comment by lokeshchdhry** [comment]
Waiting for CR to merge.

File: android/runtime/common/src/java/org/appcelerator/kroll/KrollRuntime.java [hunk 1]
```diff
@@ -541,6 +558,59 @@ public static void dispatchException(final String title, final String message, f
...
+	private static void onDisposing(KrollRuntime runtime)
+	{
+		// Validate.
+		if (runtime == null) {
+			return;
+		}
+
+		// Create a shallow copy of all listeners to be iterated down below.
+		// We do this because an invoked listener can add/remove listeners in main collection.
+		ArrayList<OnDisposingListener> clonedListeners = null;
```
Comment (garymathews) [inline android/runtime/common/src/java/org/appcelerator/kroll/KrollRuntime.java]: Why not assign `cloudListeners` right away?
```Java
ArrayList<OnDisposingListener> clonedListeners = (ArrayList<OnDisposingListener>) KrollRuntime.disposingListeners.clone();
```
I know the line is pretty long, but our linter will format it

**Review by garymathews — APPROVED** [review]
CR: **PASS**

**Pull request merged**